// @c8y/client/lib/esm2015/src/tenant-security-options/TenantSecurityOptionsService.js
import { __awaiter } from "tslib";
import { Service } from '../core';
/**
 * @description
 * This service allows to get tenant security option.
 */
export class TenantSecurityOptionsService extends Service {
    constructor() {
        super(...arguments);
        this.baseUrl = 'tenant';
        this.listUrl = 'security-options';
        this.category = 'password';
        this.propertyName = 'options';
    }
    /**
     * Get a specific tenant's security option.
     *
     * @param {ITenantSecurityOption} Tenant's security option object with key value.
     *
     * @returns Returns promise object that is resolved with the ITenantSecurityOption wrapped by IResult.
     *
     * **Example**
     * ```typescript
     *
     *     const option: ITenantSecurityOption = {
     *        key: 'limit.validity'
     *      };
     *
     *    (async () => {
     *      const {data, res} = await tenantSecurityOptionsService.detail(option);
     *   })();
     * ```
     */
    detail(tenantSecurityOption) {
        const _super = Object.create(null, {
            detail: { get: () => super.detail }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.detail.call(this, tenantSecurityOption);
        });
    }
    getDetailUrl(tenantSecurityOption) {
        return `${this.listUrl}/${this.category}/${tenantSecurityOption.key}`;
    }
    onBeforeCreate(obj) {
        return obj;
    }
}
